whelktool/scripts/cleanups/2021/04/lxl-1931-merge-series-membership.groovy

```diff
@@ -0,0 +1,256 @@
+/**
+ * Merge split SeriesMembership entities.
+ *
+ * Match on properties that correspond to marc subfields $a $v $x in 490 and 830 respectively
+ * $a have to match, plus either $v or $x given that the other also match or is non-existent
+ * We compare only strings
+ *
+ * $a is considered a match if all words match (although we allow minor differences when comparing words)
+ * $v if all numeric value match
+ * $x if all numeric value match
+ *
+ * See LXL-1931 for more info
+ *
+ */
+
+import java.text.Normalizer
+
+PrintWriter multipleMatches = getReportWriter("multiple-matches.txt")
+
+String where = "collection = 'bib' AND data#>'{@graph,1,seriesMembership,1}' IS NOT NULL"
+
+selectBySqlWhere(where) { data ->
+    String id = data.doc.shortId
+
+    List seriesMembership = data.graph[1]["seriesMembership"]
+
+    // Skip if any of the concerned properties have multiple values, we want only 1-1 comparison
+    if (seriesMembership.any { multipleValues(it) })
+        return
+
+    List from490 = []
+    List from830 = []
+
+    seriesMembership.each {
+        it.containsKey("seriesStatement") ? from490 << it : from830 << it
+    }
+
+    if (from490.isEmpty() || from830.isEmpty())
+        return
+
+
+    // There can be multiple matches, keep track of these
+    Map matchCount = seriesMembership.collectEntries { [it, 0] }
+    List matchedPairs = []
+
+    from490.each { Map sm490 ->
+        String a490 = asList(sm490["seriesStatement"])[0]
+        String v490 = asList(sm490["seriesEnumeration"])[0]
+        String x490 = sm490.inSeries?.identifiedBy?.find { it["@type"] == "ISSN" }?.value
+
+        from830.eachWithIndex { Map sm830, int idx ->
+            // Always only one Title. mainTitle sometimes (rarely) a list.
+            String a830 = asList(asList(sm830.inSeries?.instanceOf)[0]?.hasTitle?.find { it["@type"] == "Title" }?.mainTitle)[0]
+            String v830 = asList(sm830["seriesEnumeration"])[0]
+            String x830 = sm830.inSeries?.identifiedBy?.find { it["@type"] == "ISSN" }?.value
+
+            boolean aMatched = a490 && a830 ? aMatch(a490, a830) : null
+            boolean vMatched = v490 && v830 ? vMatch(v490, v830) : null
+            boolean xMatched = x490 && x830 ? xMatch(x490, x830) : null
+
+            List accepted = [[true, true, true], [true, null, true], [true, true, null]]
+
+            if ([aMatched, vMatched, xMatched] in accepted) {
+                matchedPairs << [sm490, sm830]
+                matchCount[sm490] += 1
+                matchCount[sm830] += 1
+            }
+        }
+    }
+
+    List validPairs = matchedPairs.findAll {
+        matchCount[it[0]] == 1 && matchCount[it[1]] == 1
+    }
+
+    // Remove matched objects and add the merged pair
+    validPairs.each {
+        seriesMembership.remove(it[0])
+        seriesMembership.remove(it[1])
+        seriesMembership << it[0] + it[1]
```

Comment:
Is this the form we want for the merged SeriesMembership?

Reply:
Structurally I think it should be ok. Since we exclude all properties having more than one value, we are sure to not overwrite any additional values upon merging. Although I'm not sure whether certain values that have a "better" syntax should take precedence upon conflicts.